orbit/pkg/table/mdm/mdm_windows.go

```diff
@@ -98,38 +147,111 @@ func Generate(ctx context.Context, queryContext table.QueryContext) ([]map[strin
+			return nil, fmt.Errorf("mdm command response parsing: %s ", err)
+		}
+
 		return []map[string]string{
 			{
-				"enrollment_status":  deviceEnrollmentStatus,
-				"enrolled_user":      enrollmentURI,
-				"mdm_command_input":  inputCmd,
-				"mdm_command_output": outputCmd,
+				"enrollment_status":      deviceEnrollmentStatus,
+				"enrolled_user":          enrollmentURI,
+				"mdm_command_input":      inputCmd,
+				"mdm_command_output":     minimalOutputCmd,
+				"raw_mdm_command_output": outputCmd,
 			},
 		}, nil
 	}
 
 	// returning table enrollment status + cmd output status if present
 	return []map[string]string{
 		{
-			"enrollment_status": deviceEnrollmentStatus,
-			"enrolled_user":     enrollmentURI,
-			"mdm_command_input": "",
-			"mdm_command_ouput": "",
+			"enrollment_status":      deviceEnrollmentStatus,
+			"enrolled_user":          enrollmentURI,
+			"mdm_command_input":      "",
+			"mdm_command_ouput":      "",
+			"raw_mdm_command_output": "",
 		},
 	}, nil
 }
 
-// builtin utf16tostring string expects a uint16 array but we have a pointer to a uint16
-// so we need to cast it after converting it to an unsafe pointer
-// this is a common pattern though the buffer size varies
-// see https://golang.org/pkg/unsafe/#Pointer for more details
+// dummy charset reader just to satisfy the xml decoder requirements
+func identReader(encoding string, input io.Reader) (io.Reader, error) {
+	return input, nil
+}
+
+// getCommandResponseData returns the response data for a given command
+func getCmdResponseData(outputCmd string) (string, error) {
+	var responseData string
+
+	// creating a new SyncML message object
+	messageObject := new(SyncMLMessage)
+
+	// parsing output SyncML message
+	d := xml.NewDecoder(bytes.NewReader([]byte(outputCmd)))
+	d.CharsetReader = identReader
+
+	// decoding the XML message
+	if err := d.Decode(messageObject); err != nil {
+		return "", err
+	}
+
+	// getting response data from output message
+	if len(messageObject.Body.Item) > 0 {
+		for _, element := range messageObject.Body.Item {
+
+			// getting the results tag for the input commands
+			if element.XMLName.Local != "Results" {
+				continue
+			}
+
+			// results will be appended in a comma separated list
+			if len(element.Item) > 0 {
+
+				// extracting the data from the result
+				workStr := element.Item[0].Data
+				if len(workStr) == 0 {
+					workStr = "data_not_set" // default value for empty data
+				}
+				responseData += workStr
+			}
+		}
+	}
+
+	return responseData, nil
+}
+
+// builtin windows.UTF16ToString string expects a uint16 array but we have a uint16 pointer
+// so we are allocating dynamic memory and moving data around before calling windows.UTF16ToString
 func localUTF16toString(ptr unsafe.Pointer) (string, error) {
 	if ptr == nil {
 		return "", errors.New("failed UTF16 conversion due to null pointer")
 	}
 
-	uint16ptrarr := (*[maxBufSize]uint16)(ptr)[:]
-	return windows.UTF16ToString(uint16ptrarr), nil
```

Comment:
This was not working right? Or was it working but unreliably?

Reply:
it indeed worked fine, but I feel better controlling exactly how memory is laid out using native API calls instead of relying on this language hack to do the same. the old code is being used to do this trick on windows - see [here](https://github.com/carmo-evan/libman/blob/0613a2a22d7466c2123faee6bb396a9e2fdd8098/userdir/ud_windows.go#L61)